File: AngularProject/app/service/applicationService.js
const application = angular.module('application', [])
application.factory('applicationService', ['$http', '$q', '$window', function ($http, $q, $window) {
    return {
        createApplication: function (application, reasons, office) {
            const deferred = $q.defer()
            const requestBody = {
                application: {
                    additionalInfo: application.additionalInfo,
                    educationalProgram: application.educationalProgram,
                    groups: application.groups,
                    reasons: reasons
                },
                office: office
            }
            $http.post('http://localhost:8080/application/createApplication', requestBody).then(function success(response) {
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        getApplications: function () {
            const deferred = $q.defer()
            $http.get('http://localhost:8080/application/getApplications').then(function success(response) {
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        getProviderApplications: function () {
            const deferred = $q.defer()
            const login = localStorage.getItem('UserLogin')
            $http.get(`http://localhost:8080/application/getProviderApplications?login=${login}`).then(function success(response) {
                deferred.resolve(response.data)
            })
            return deferred.promise
        },

        getEducationalProgramById: function (id) {
            const deferred = $q.defer()
            const login = localStorage.getItem('UserLogin')
            $http.get(`http://localhost:8080/application/getEducationalProgramById?id=${id}&login=${login}`).then(function success(response) {
                if (response.data.title) {
                    alert(response.data.title)
                    $window.location.href = '#!/application/reviewers/view'
                }
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        getApplication: function (id) {
            const deferred = $q.defer()
            const login = localStorage.getItem('UserLogin')
            $http.get(`http://localhost:8080/application/getApplication?id=${id}&login=${login}`).then(function success(response) {
                if (response.data.title) {
                    alert(response.data.title)
                    $window.location.href = '#!/application/reviewers/view'
                }
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        getHistory: function (id) {
            const deferred = $q.defer()
            $http.get(`http://localhost:8080/application/getHistory?id=${id}`).then(function success(response) {
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        finalDecision: function (id, decision) {
            const deferred = $q.defer()
            $http.post(`http://localhost:8080/application/finalDecision?id=${id}&decision=${decision}`).then(function success(response) {
                deferred.resolve(response.data)
            })
            return deferred.promise
        },
        rejectAndAccept: function (id, application, status, section) {
            const deferred = $q.defer()
            const requestBody = {
                application: application,
                login: localStorage.getItem('UserLogin'),
                status: status,
                typeOfSection: section
            }
            $http.post(`http://localhost:8080/application/rejectAndAccept`, requestBody).then(function success(response) {
                if (response.data.title !== 'ok!') {
                    alert(response.data.title)
                    $window.location.href = '#!/application/reviewers/view'
                }
                deferred.resolve(response.data)
            })
            return deferred.promise
        }
    }
}])

File: spring-project/src/main/java/by/project/first/service/ApplicationService.java
package by.project.first.service;

import by.project.first.controllers.ReqAndRes.ApplicationCreateRequest;
import by.project.first.controllers.ReqAndRes.RegularReviewerResponse;
import by.project.first.controllers.ReqAndRes.RejectAndAcceptRequest;
import by.project.first.models.ApplicationModels.ApplicationModel;
import by.project.first.models.ApplicationModels.GroupsModel;
import by.project.first.models.ApplicationModels.ReasonsModel;
import by.project.first.models.ApplicationModels.ResponseToApplicationModel;
import by.project.first.models.ApplicationModels.WorkerModelForResponse;
import by.project.first.models.Message;
import by.project.first.models.OfficeModel;
import by.project.first.models.UserModel;
import by.project.first.repositories.ApplicationRepo;
import by.project.first.repositories.GroupRepo;
import by.project.first.repositories.OfficeRepo;
import by.project.first.repositories.ReasonRepo;
import by.project.first.repositories.ResponseToApplicationRepo;
import by.project.first.repositories.UserRepo;
import by.project.first.repositories.WorkerRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

@Service
public class ApplicationService {

    private final OfficeRepo officeRepo;
    private final ApplicationRepo applicationRepo;
    private final ResponseToApplicationRepo responseToApplicationRepo;
    private final WorkerRepo workerRepo;
    private final ReasonRepo reasonRepo;
    private final UserRepo userRepo;
    private final GroupRepo groupRepo;

    @Autowired
    public ApplicationService(OfficeRepo officeRepo, ApplicationRepo applicationRepo, WorkerRepo workerRepo,
                              ReasonRepo reasonRepo, UserRepo userRepo, GroupRepo groupRepo,
                              ResponseToApplicationRepo responseToApplicationRepo) {
        this.officeRepo = officeRepo;
        this.applicationRepo = applicationRepo;
        this.workerRepo = workerRepo;
        this.reasonRepo = reasonRepo;
        this.userRepo = userRepo;
        this.groupRepo = groupRepo;
        this.responseToApplicationRepo = responseToApplicationRepo;
    }

    public Message createApplication(ApplicationCreateRequest request) {
        OfficeModel office = officeRepo.findByName(request.getOffice().getName());
        Set<ReasonsModel> reasons = request.getApplication().getReasons();
        Set<GroupsModel> groups = request.getApplication().getGroups();

        reasonRepo.saveAll(reasons);
        groupRepo.saveAll(groups);
        saveOfficeAndApplication(office, reasons, groups,
                request.getApplication().getAdditionalInfo(), request.getApplication().getEducationalProgram());

        return new Message("ok!");
    }

    public void saveOfficeAndApplication(OfficeModel office, Set<ReasonsModel> reasons, Set<GroupsModel> groups,
                                         String additionalInfo, String educationalProgram) {
        ApplicationModel newApplication = applicationRepo.save(new ApplicationModel(reasons, groups,
                educationalProgram, additionalInfo, office.getName(), office.getLocation())
        );

        office.setLastApplication(newApplication);
        office.getOfficeApplications().add(newApplication);
        officeRepo.save(office);
    }

    public Set<ApplicationModel> getApplications() {
        Set<ApplicationModel> notAnsweredApplications = new HashSet<>();
        applicationRepo.findAll().forEach(application -> {
            if (application.getStatus().equals("WAIT_FOR_AN_ANSWER")) {
                notAnsweredApplications.add(application);
            }
        });
        return notAnsweredApplications;
    }

    public RegularReviewerResponse getEducationalProgram(Long id, String login) {
        return getResponsesAndApplication(login, id);
    }

    public RegularReviewerResponse getResponsesAndApplication(String login, Long id) {
        UserModel user = userRepo.findByLogin(login);
        return new RegularReviewerResponse(applicationRepo.findById(id).get(),
                responseToApplicationRepo.findAllByUser(user)
        );
    }

    public RegularReviewerResponse getApplication(Long id, String login) {
        RegularReviewerResponse response = getResponsesAndApplication(login, id);
        Set<WorkerModelForResponse> workers = new HashSet<>();
        response.getApplication().getReasons().forEach(reason ->
                workers.add(new WorkerModelForResponse(
                        workerRepo.findById(reason.getWorkerID()).get(), reason.getReason()))
        );
        return new RegularReviewerResponse(workers, response.getApplication(), response.getResponses());
    }

    public Message rejectAndAccept(RejectAndAcceptRequest request) {
        UserModel user = userRepo.findByLogin(request.getLogin());
        ResponseToApplicationModel response = new ResponseToApplicationModel(request.getStatus(),
                request.getTypeOfSection(), user, request.getApplication(), new Date()
        );
        responseToApplicationRepo.save(response);
        return new Message("ok!");
    }

    public Iterable<ResponseToApplicationModel> getHistory(Long id) {
        return responseToApplicationRepo.findAllByApplicationID(applicationRepo.findById(id).get());
    }

    public Set<ApplicationModel> getProviderApplications(String login) {
        Set<ApplicationModel> providerApplications = new HashSet<>();
        officeRepo.findAllByLeaderID(userRepo.findByLogin(login)).forEach(office ->
                providerApplications.addAll(office.getOfficeApplications())
        );
        return providerApplications;
    }

    public ResponseEntity<Message> finalDecision(Long id, String decision) {
        Optional<ApplicationModel> application = applicationRepo.findById(id);
        OfficeModel office = officeRepo.findByLastApplication(application.get());

        if (office == null) return ResponseEntity.status(400).body(new Message("Can not find office!"));
        if (decision.equals("ACCEPT")) return finalDecisionAccept(office, application.get(), decision);
        if (decision.equals("REJECT")) return finalDecisionReject(application.get(), decision);

        return ResponseEntity.status(400).body(new Message("Something wrong"));
    }

    public ResponseEntity<Message> finalDecisionAccept(OfficeModel office,
                                                       ApplicationModel application, String decision) {
        office.setDateOfLastPermission(new Date());
        officeRepo.save(office);
        application.setStatus(decision);
        applicationRepo.save(application);
        return ResponseEntity.ok(new Message("ok!"));
    }

    public ResponseEntity<Message> finalDecisionReject(ApplicationModel application, String decision) {
        application.setStatus(decision);
        applicationRepo.save(application);
        return ResponseEntity.ok(new Message("ok!"));
    }

}


